File: backend/src/main/java/com/andrew/hdss/models/FormResponse.java
package com.andrew.hdss.models;

import com.andrew.hdss.models.enums.FormResponseStatus;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.Instant;
import java.time.LocalDateTime;

@Entity
@Table(name = "form_responses")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class FormResponse {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(unique = true, nullable = false)
    private String clientId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "visit_id", nullable = false)
    private Visit visit;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "form_id", nullable = false)
    private Form form;

    @Column(nullable = false)
    private Integer formVersion;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private FormResponseStatus status;

    @Column(nullable = false)
    private LocalDateTime startedAt;

    private LocalDateTime completedAt;
}

File: android/app/src/main/java/com/andrew/hdss/data/models/FormResponse.kt
package com.andrew.hdss.data.models

import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.PrimaryKey
import com.andrew.hdss.data.dtos.FormResponsePushDto
import com.andrew.hdss.data.models.enums.FormResponseStatus
import java.time.LocalDateTime

@Entity(
    tableName = "form_responses",
    foreignKeys = [
        ForeignKey(
            entity = Visit::class,
            parentColumns = ["id"],
            childColumns = ["visitId"],
            onDelete = ForeignKey.CASCADE
        ),
        ForeignKey(
            entity = Form::class,
            parentColumns = ["id"],
            childColumns = ["formId"],
            onDelete = ForeignKey.CASCADE
        )
    ]
)
data class FormResponse(
    @PrimaryKey
    val id: String,
    val visitId: String,
    val formId: Long,
    val formVersion: Int,
    val status: FormResponseStatus,
    val startedAt: LocalDateTime,
    val completedAt: LocalDateTime,
    val synced: Boolean = false
)

fun List<FormResponse>.toPushDtos(): List<FormResponsePushDto>{
    return map{
        formResponse ->
        FormResponsePushDto(
            clientId = formResponse.id,
            visitClientId = formResponse.visitId,
            formId = formResponse.formId,
            formVersion = formResponse.formVersion,
            status = formResponse.status,
            startedAt = formResponse.startedAt.toString(),
            completedAt = formResponse.completedAt.toString()
        )
    }
}